### PR TITLE
Add ability to perform or not scenarios depending on the previous res…

### PR DESCRIPTION
**Pull request by Yauhenda**
…ults

### CODE REVIEW

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 1]
```diff
@@ -941,7 +945,39 @@ private void performStorySteps(RunContext context, List<PerformableSteps> storyS
 
         private void performScenarios(RunContext context) throws InterruptedException {
             for (PerformableScenario scenario : scenarios) {
+                Meta meta = scenario.getScenario().getMeta();
+                if (meta.hasProperty(DEPENDS_ON_ID)) {
+                    String dependsOnId = meta.getProperty(DEPENDS_ON_ID);
+                    Boolean dependsOnSuccess = successfulScenariosId.get(dependsOnId);
```
[the remark this answers is not in the document]
Comment (Yauhenda) [reply jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]: Changed to `performDependentScenario` to keep existing logic.

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 2]
```diff
@@ -940,8 +944,55 @@ private void performStorySteps(RunContext context, List<PerformableSteps> storyS
...
+                        Throwable cause = performDependentScenario == null ? new DependsOnNotFoundException(dependsOnId)
+                                : new DependsOnFailedException(dependsOnId);
+                        UUIDExceptionWrapper wrappedCause = new UUIDExceptionWrapper(cause);
+                        Scenario scenario = performableScenario.getScenario();
+                        beforeScenario(context, scenario);
+                        if (performableScenario.hasExamples()) {
+                            context.reporter().beforeExamples(scenario.getSteps(), scenario.getExamplesTable());
+                            for (ExamplePerformableScenario exampleScenario : performableScenario.exampleScenarios) {
+                                PerformableSteps steps = exampleScenario.steps;
+                                List<Step> stepList = steps.steps;
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]: AFAIK, there also should be invocations of `StoryReporter#example`, please, review all reporting flow thoroughly

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 3]
```diff
@@ -940,8 +944,55 @@ private void performStorySteps(RunContext context, List<PerformableSteps> storyS
...
+                Meta meta = performableScenario.getScenario().getMeta();
+                if (meta.hasProperty(DEPENDS_ON_ID)) {
+                    String dependsOnId = meta.getProperty(DEPENDS_ON_ID);
+                    Boolean performDependentScenario = successfulScenariosId.get(dependsOnId);
+                    if (performDependentScenario == null || !performDependentScenario) {
+                        Throwable cause = performDependentScenario == null ? new DependsOnNotFoundException(dependsOnId)
+                                : new DependsOnFailedException(dependsOnId);
+                        UUIDExceptionWrapper wrappedCause = new UUIDExceptionWrapper(cause);
+                        Scenario scenario = performableScenario.getScenario();
+                        beforeScenario(context, scenario);
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]: also one more thing: how are GovenStories of scenario-level reported?

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 4]
```diff
@@ -940,8 +944,55 @@ private void performStorySteps(RunContext context, List<PerformableSteps> storyS
         }
 
         private void performScenarios(RunContext context) throws InterruptedException {
-            for (PerformableScenario scenario : scenarios) {
-                scenario.perform(context);
+            for (PerformableScenario performableScenario : scenarios) {
+                Meta meta = performableScenario.getScenario().getMeta();
+                if (meta.hasProperty(DEPENDS_ON_ID)) {
```
Comment (ikalinin1) [inline jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]: Could it be refactored to something like this?
```java
if (isDependant(meta) && isPerformDependant(meta)) {
throwExecptionTo...();
failScenario();
}
```

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 5]
```diff
@@ -819,19 +822,21 @@ public PerformableStory get(Story story) {
     }
 
     private static void performGivenStories(RunContext context, List<PerformableStory> performableGivenStories,
-            GivenStories givenStories) throws InterruptedException {
+                                            GivenStories givenStories, Optional<UUIDExceptionWrapper> causeToSkip)
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]: formatting is a bit awkward

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 6]
```diff
@@ -930,18 +940,43 @@ public void reportFailures(FailureContext context) {
             }
         }
 
-        private void performStorySteps(RunContext context, List<PerformableSteps> storySteps, Stage stage)
+        private void performStorySteps(RunContext context, List<PerformableSteps> storySteps, Stage stage,
+                                       Optional<UUIDExceptionWrapper> causeToSkip)
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]: formatting

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 7]
```diff
@@ -930,18 +940,43 @@ public void reportFailures(FailureContext context) {
...
+            if (meta.hasProperty(DEPENDS_ON_ID)) {
+                String dependsOnId = meta.getProperty(DEPENDS_ON_ID);
+                Boolean performDependentScenario = successfulScenariosId.get(dependsOnId);
+                if (performDependentScenario == null || !performDependentScenario) {
+                    Throwable cause = performDependentScenario == null ? new DependsOnNotFoundException(dependsOnId)
+                            : new DependsOnFailedException(dependsOnId);
+                    wrappedCause = new UUIDExceptionWrapper(cause);
+                }
+            }
+            return Optional.ofNullable(wrappedCause);
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]:
```java
Meta meta = scenario.getScenario().getMeta();
if (meta.hasProperty(DEPENDS_ON_ID)) {
    String dependsOnId = meta.getProperty(DEPENDS_ON_ID);
    Boolean performDependentScenario = successfulScenariosId.get(dependsOnId);
    if (performDependentScenario == null) {
        return Optional.of(new UUIDExceptionWrapper(new DependsOnNotFoundException(dependsOnId)));
    }
    if (!performDependentScenario) {
        return Optional.of(new UUIDExceptionWrapper(new DependsOnFailedException(dependsOnId)));
    }
}
return Optional.empty();
```
`Optional.of(new UUIDExceptionWrapper(...))` - can be moved to a separate method

File: jbehave-core/src/test/java/org/jbehave/core/embedder/PerformableTreeBehaviour.java [hunk 8]
```diff
@@ -27,18 +31,19 @@
...
+import org.jbehave.core.exception.DependsOnFailedException;
+import org.jbehave.core.exception.DependsOnNotFoundException;
 import org.jbehave.core.failures.BatchFailures;
+import org.jbehave.core.failures.UUIDExceptionWrapper;
 import org.jbehave.core.io.StoryLoader;
-import org.jbehave.core.model.GivenStories;
-import org.jbehave.core.model.Meta;
-import org.jbehave.core.model.Scenario;
-import org.jbehave.core.model.Story;
+import org.jbehave.core.model.*;
```
Comment (valfirst) [inline jbehave-core/src/test/java/org/jbehave/core/embedder/PerformableTreeBehaviour.java]: do not use star-imports

File: jbehave-core/src/test/java/org/jbehave/core/embedder/PerformableTreeBehaviour.java [hunk 9]
```diff
@@ -106,6 +111,138 @@ public void shouldNotSkipStoryWhenGivenStoryIsPassed() {
...
+        EmbedderMonitor embedderMonitor = mock(EmbedderMonitor.class);
+        BatchFailures failures = mock(BatchFailures.class);
+        StoryReporter storyReporter = mock(StoryReporter.class);
+        when(configuration.storyReporter(STORY_PATH)).thenReturn(storyReporter);
+        PerformableTree performableTree = new PerformableTree();
+        RunContext runContext = performableTree.newRunContext(configuration, candidateSteps,
+                embedderMonitor, new MetaFilter(), failures);
+        performableTree.addStories(runContext, Collections.singletonList(story));
+        performableTree.perform(runContext, story);
+        verify(storyReporter, never()).failed(any(), any());
```
Comment (valfirst) [inline jbehave-core/src/test/java/org/jbehave/core/embedder/PerformableTreeBehaviour.java]: I believe you need to add more tests to cover all various cases, e.g. there is no test for GivenStories

File: jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java [hunk 10]
```diff
@@ -895,6 +901,10 @@ public void add(PerformableScenario performableScenario) {
 
         @Override
         public void perform(RunContext context) throws InterruptedException {
+            perform(context, Optional.empty());
+        }
+
+        public void perform(RunContext context, Optional<UUIDExceptionWrapper> causeToSkip) throws InterruptedException {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/embedder/PerformableTree.java]: is it required to keep this and all similar methods public/protected?

**Pull request closed without merging**